### ISSUE TITLE
Feature Request: managers

### ISSUE THREAD

**Issue by RWOverdijk** [balderdashy/sails#2178]
Good people :)

Using `sails.service.someservice.somemethod()` is starting to annoy me. I'm not going to use globals... So what's left is `require().` That's also not ideal. So I propose creating managers.

I want to propose splitting the services, models and controllers (as services) up and setting them on managers. Before I start writing up the entire proposal, I want to first give you the gist of it, and then see if you're interested. The concept, and implementation are loosely based on the ZF2 implementation, by the way. A strict copy wouldn't take advantage of node.js :)

```javascript
var serviceManager = require('service-manager');

// Add a service
serviceManager.add('user', userService);

// Get a service
serviceManager.get('user', userService);

// Factories
serviceManager.add('user', function(serviceManager) {
  // Dependency injection
  userService.setAccount(serviceManager.get('account'));

  return userService;
});
```

I think you get the gist... Same goes for Models, [calling instances entities](https://github.com/balderdashy/waterline/issues/547) and creating an EntityManager.

**Update:**

The main advantage I see with this, is that it's more clear what you're fetching and why. You put the tasks in their place. You can configure services, set up DI through factories and more cool stuffs.

**Comment by elennaro** [comment]
+1

**Comment by sailsbot** [comment]
Thanks for posting, @RWOverdijk.  I'm a repo bot-- nice to meet you!

It has been 60 days since there have been any updates or new comments on this page.  If this issue has been resolved, feel free to disregard the rest of this message.  On the other hand, if you are still waiting on a patch, please:

  + review our [contribution guide](https://github.com/balderdashy/sails/blob/master/CONTRIBUTING.md) to make sure this submission meets our criteria (only _verified bugs_ with documented features, please;  no questions, commentary, or bug reports about undocumented features or unofficial plugins)
  + create a new issue with the latest information, including updated version details with error messages, failing tests, and a link back to [the original issue]().  This allows GitHub to automatically create a back-reference for future visitors arriving from search engines.

Thanks so much for your help!

**Comment by RWOverdijk** [comment]
Soooo, no feedback on this?

**Comment by simonblee** [comment]
+1

**Comment by mikermcneil** [comment]
@RWOverdijk I see this as a broader issue beyond Sails and with dep. management in general.  Personally, I'm a big fan of CJS, but I know that it's not everyone's favorite.

**Comment by RWOverdijk** [comment]
@mikermcneil But specifically for sails it could help a lot to use the service pattern. Right now, I have to loop through my services on bootstrap to call "init" because services are loaded too early.

**Comment by mikermcneil** [comment]
Here's an example of how you might factor that out as a hook:

```
module.exports = function ServicesThatHaveInitFunctionsHook (sails) {
  return  {
    initialize: function (cb) {
      // Load relevant services into sails from the api/services folder
      // -- or wait until after the normal services hook finishes using `sails.after()`
      // and grab them from sails.services that way using e.g. `_.filter` to avoid the extra disk read
      // (although under normal circumstances the require cache is hot at this point so it wouldn't matter)
      var servicesWithInitFns = [];
      // TODO
      
       // Run `init` on each
       async.each(servicesWithInitFns, function eachService(serviceDef, next) {
         serviceDef.init(next);
       }, function afterwards(err) {  return cb(err);  })

       // Continue loading sails
       return cb();
    }
  };
};
```

But honestly, better yet, I'd recommend just writing those services using `init` functions as hooks (using the built-in `initialize` function instead).  That's pretty much exactly the problem hooks and events have solved for Sails core from the beginning, so I think you might find it useful for that sort of stuff.  Just my 2¢.

**Comment by RWOverdijk** [comment]
Sounds good to me :) Hope this _does_ work with the reload hook
